### PR TITLE
Global exception logger

### PR DESCRIPTION
**Pull request by 84634E1A607A**
**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by 84634E1A607A** [comment]
Crash report description needs polishing. (Feel free to change it)

**Comment by hhyyrylainen** [comment]
I'm 90% certain this will not change anything as only native code crashes (which is what Breakpad detects) are reportable by the launcher. There's an issue for the launcher to make it detect exception stack traces and offer to report them from logs: https://github.com/Revolutionary-Games/Thrive-Launcher/issues/163
Looks like I maybe had forgotten to open an issue for that in the launcher, may have just been on my personal TODO list, so I had to open one now.

File: src/engine/StartupActions.cs [hunk 1]
```diff
@@ -15,6 +16,13 @@ private StartupActions()
         // TODO: for devbuilds it would be nice to print the hash here
         GD.Print("This is Thrive version: ", Constants.Version);
 
+        // Add crash logger if debugger not attached
+        if (!Debugger.IsAttached)
+        {
+            GD.Print("Crash logger attached");
```
Comment (hhyyrylainen) [inline src/engine/StartupActions.cs:22]: This is not really a crash logger, I'd name this unhandled exception logger, because as I said in my comment, I think none of the reports that players have reported can have been like this. So I'd like to reword this to not confuse players reporting these with the native crashes. So I'd rename all instances of "crash" in this PR to be "Unhandled Exception"

File: src/engine/CrashLogger.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+﻿using Godot;
+
+/// <summary>
+///   Crash logger for any unhandled Mono exceptions
+/// </summary>
+public class CrashLogger
+{
+    public static void UnhandledException(object sender, UnhandledExceptionArgs args)
+    {
+        GD.PrintErr("------------ Begin of Crash Log ------------\n" +
```
Comment (hhyyrylainen) [inline src/engine/CrashLogger.cs:10]: Don't concatenate multiple strings with +, it's very inefficient in C#. Instead I think this could use just commas to separate the parts, if you don't want to use a string builder here.

**Review by hhyyrylainen — APPROVED** [review]
I did a few slight tweaks myself and tested this. Works as intended I think. In the future we could maybe do something like exiting to a separate scene allowing viewing and copying the exception log to make it easier to report these.

**Comment by 84634E1A607A** [comment]
> In the future we could maybe do something like exiting to a separate scene

I don't think it is possible, because an unhandled exception leaves the logger little time, as long as the logger function returns, game crashes.

**Comment by hhyyrylainen** [comment]
> I don't think it is possible, because an unhandled exception leaves the logger little time, as long as the logger function returns, game crashes.

Oh yeah, I suppose that is true... Perhaps there's a way to tell Godot to not actually close due to the exception? Or maybe we could write the exception to a specifically named file, which the launcher could then detect and offer to submit it as a crash report?
